library/core/src/hint.rs

```diff
@@ -175,34 +175,27 @@ pub fn spin_loop() {
         unsafe { crate::arch::x86_64::_mm_pause() };
     }
 
-    // RISC-V platform spin loop hint implementation
+    #[cfg(target_arch = "riscv32")]
     {
-        // RISC-V RV32 and RV64 share the same PAUSE instruction, but they are located in different
-        // modules in `core::arch`.
-        // In this case, here we call `pause` function in each core arch module.
-        #[cfg(target_arch = "riscv32")]
-        {
-            crate::arch::riscv32::pause();
-        }
-        #[cfg(target_arch = "riscv64")]
-        {
-            crate::arch::riscv64::pause();
-        }
+        crate::arch::riscv32::pause();
     }
 
-    #[cfg(any(target_arch = "aarch64", all(target_arch = "arm", target_feature = "v6")))]
+    #[cfg(target_arch = "riscv64")]
     {
-        #[cfg(target_arch = "aarch64")]
-        {
-            // SAFETY: the `cfg` attr ensures that we only execute this on aarch64 targets.
-            unsafe { crate::arch::aarch64::__isb(crate::arch::aarch64::SY) };
-        }
-        #[cfg(target_arch = "arm")]
-        {
-            // SAFETY: the `cfg` attr ensures that we only execute this on arm targets
-            // with support for the v6 feature.
-            unsafe { crate::arch::arm::__yield() };
-        }
+        crate::arch::riscv64::pause();
+    }
+
+    #[cfg(target_arch = "aarch64")]
+    {
+        // SAFETY: the `cfg` attr ensures that we only execute this on aarch64 targets.
+        unsafe { crate::arch::aarch64::__isb(crate::arch::aarch64::SY) };
+    }
+
+    #[cfg(all(target_arch = "arm", target_feature = "v6"))]
+    {
+        // SAFETY: the `cfg` attr ensures that we only execute this on arm targets
+        // with support for the v6 feature.
```

Comment:
I'm not sure those (or any of the above) comments make sense. As far as I can tell, all the functions in `core::arch::...` are cfg-ed anyway, so we would not be able to call them on a wrong arch. Should I delete them?

Reply:
Functions in `core::arch::...` can be called just fine if the relevant `cfg(target_feature = "...")` is not true at compile time. In that case you need to do runtime detection to check if it is supported before calling however.

Reply:
I'm not sure how this could be true.

`__yield` is defined in `arm_shared`. `arm_shared` is marked as `#[cfg(any(target_arch = "arm", target_arch = "aarch64", doc))]`. `__yield` itself is marked as `#[cfg(any(target_feature = "v6", target_arch = "aarch64", doc))]`.

So this results in `__yield` being available if `cfg(all(any(target_arch = "arm", target_arch = "aarch64", doc), any(target_feature = "v6", target_arch = "aarch64", doc)))`. `doc` and `target_arch = "aarch64"` are not true in this context, so we are left with `cfg(all(any(target_arch = "arm"), any(target_feature = "v6")))`, which can be simplified to `cfg(all(target_arch = "arm", target_feature = "v6"))`.

Unless I'm missing something, this is exactly the same as the `cfg` on this block?...

Reply:
Looks like you are right in this case. For x86 intrinsics this is definitively not the case though. Those are never `cfg`ed on target features.